components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java

```diff
@@ -358,11 +358,66 @@ public void deleteGatewayArtifacts(String apiId) throws APIManagementException {
         return apiRuntimeArtifactDtoList;
     }
 
+    public List<APIRuntimeArtifactDto> retrieveGatewayArtifactsByAPIIDs(List<String> apiIds, String[] labels,
+                                                                        String tenantDomain)
+            throws APIManagementException {
+
+        String query = SQLConstants.RETRIEVE_ARTIFACTS_BY_APIID_AND_LABEL;
+        query = query.replaceAll(SQLConstants.GATEWAY_LABEL_REGEX, String.join(",",Collections.nCopies(labels.length, "?")));
+        List<APIRuntimeArtifactDto> apiRuntimeArtifactDtoList = new ArrayList<>();
+        try (Connection connection = GatewayArtifactsMgtDBUtil.getArtifactSynchronizerConnection();
+             PreparedStatement preparedStatement = connection.prepareStatement(query)) {
+            for (String apiId: apiIds) {
+                preparedStatement.setString(1, apiId);
```

Comment:
If I understand correctly, from this, we are sending one query per each APIID?
Here the number of APIIDs can be high like 100, 1000 2000 etc right.. we are looking into those numbers right? In that case its also bit dangerous to use IN with all API IDs at once.. 
Can we split the List<String> apiIds into smaller lists (may be one containing 25 APIIDs), then when processing each smaller list we can do an IN altogether? Then we reduce the query count in 25 times..

Reply:
Fixed with 08e485e24c0c161e5fe355208c8cde2a7ceba5df